---
title: "SummarizedExperiment_generation"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Filter the m6Ads used in other projects
```{r}
m6Ads_x <- readRDS("m6Ads_multi.rds")
m6Ads <- m6Ads[,!m6Ads$Technique%in%c("MAZTER-Seq","DART-Seq","m6A-REF-Seq")]
m6Ads <- m6Ads[,-1]
write.csv( colData(m6Ads), "colData.csv")
m6Ads <- m6Ads[rowSums(assays(m6Ads)[["Model"]], na.rm = TRUE)>0,]
rowData(m6Ads) <- NULL
m6Ads$Sample_ID <- seq_len(ncol(m6Ads))
assay(m6Ads)[is.na(assay(m6Ads))] <- 0
assays(m6Ads)[["Pred"]] <- NULL
assays(m6Ads)[["Offset"]] <- NULL
metadata(m6Ads) <- list()

library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(BSgenome.Hsapiens.UCSC.hg19)

all_motifs <- RegionPropertiesFeatures::sampleSequence("DRACH", exons(TxDb.Hsapiens.UCSC.hg19.knownGene), Hsapiens)
all_motifs <- keepStandardChromosomes(all_motifs, pruning.mode = "coarse")
all_motifs <- all_motifs-2
zero_motifs <- subsetByOverlaps(all_motifs, m6Ads, invert = TRUE, type="equal")
rm(all_motifs)
SE_full <- SummarizedExperiment(assays = rbind(assay(m6Ads), matrix(0,nrow=length(zero_motifs),ncol=ncol(m6Ads))),
                     colData = colData(m6Ads),
                     rowRanges = c(rowRanges(m6Ads), zero_motifs))
SE_full <- sort(SE_full)
SE_full <- keepStandardChromosomes(SE_full, pruning.mode = "coarse")
SE_full <- SE_full[names(SE_full)!="chrM",]
```

##Create 4 types of features for SE_full
```{r}
metadata(SE_full)[["iRNA"]] <- RegionPropertiesFeatures::PredictiveSequenceFeatures(rowRanges(SE_full)+20, Hsapiens, encoding = "iRNA")

```



## Merge all of the positive and negative sites

```{r}
N <- nrow(Design_sorted)
site_lst <- list()

for (i in 1:N){
 site_lst[[i]] <- readRDS(paste0("classificationDataSet/sample_",i,".rds"))
}

rowRanges <- Reduce(function(x,y) c(x,y) ,site_lst)
rowRanges <- reduce(rowRanges, min.gapwidth=0L)
all(width(rowRanges) == 1)

##Prepare 2 assay matrices
ModelAssay <- matrix(NA, nrow = length(rowRanges), ncol = N)
for (i in 1:N){
  fol_i <- findOverlaps(rowRanges, site_lst[[i]])
  ModelAssay[queryHits(fol_i),i] <- site_lst[[i]]$Y[subjectHits(fol_i)]
}
```

## Create Row data (features)

```{r}
all_positive <- Reduce(function(x,y) c(x,y), site_lst[1:N])
all_positive <- reduce(all_positive, min.gapwidth=0L)

sample_region <- trim(reduce(all_positive+1600))
start(sample_region)[as.vector(seqnames(sample_region)=="chrM")] <- 1
end(sample_region)[as.vector(seqnames(sample_region)=="chrM")] <- 16571

all_DRACH <- RegionPropertiesFeatures::sampleSequence("DRACH", sample_region, genome_hg19)

library(phastCons100way.UCSC.hg19)
library(fitCons.UCSC.hg19)

rowData <- RegionPropertiesFeatures::RegionPropertiesFeatures(rowRanges,
                                                         transcriptdb=txdb_hg19,
                                                         sequence=genome_hg19,
                                                         gscores=c(phastCons = phastCons100way.UCSC.hg19,
                                                                   fitCons = fitCons.UCSC.hg19),
                                                         clusteringY=c(DRACH = all_DRACH))

rowData <- cbind(rowData, RegionPropertiesFeatures::PredictiveSequenceFeatures(rowRanges+20, genome_hg19))
colData <- Design_sorted
rowData <- rowData[,matrixStats::colVars(as.matrix(rowData)) != 0]
mcols(rowRanges) <- rowData

m6Ads <- SummarizedExperiment(assays = list(Model = ModelAssay),
                                           rowRanges = rowRanges,
                                           colData = colData)
saveRDS(m6Ads,"../m6Ads.rds")
```

###Add whistle features
```{r}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(BSgenome.Hsapiens.UCSC.hg19)
library(fitCons.UCSC.hg19)
library(phastCons100way.UCSC.hg19)

txdb_hg19 <- TxDb.Hsapiens.UCSC.hg19.knownGene

whistle_feature <- WhistleR::transcriptomicFeatures(
   mod_gr = rowRanges(m6Ads),
# TxDb object for the transcript annotation.
   txdb = txdb_hg19,
# BSgenome object for the genome sequence.
   bsgnm = Hsapiens,
# GScores object for the Fitness Consequence scores.
   fc = fitCons.UCSC.hg19,
# GScores object for the PhastCons scores.
   pc = phastCons100way.UCSC.hg19,
# The clustering effect on DRACH motifs.
   motif_clustering = "DRACH"
)

whistle_feature <- cbind(whistle_feature, WhistleR::sequenceFeatures(rowRanges(m6Ads)+20, Hsapiens))

metadata(m6Ads)[[1]] <- whistle_feature*1
```

###Add regulators

```{r}
library(magrittr)
#prepare the RBP binding site collection for each m6A regulator
file_names <- list.files("./regulator")
regulator_names <- file_names %>% gsub("_.*","",.) %>% unique
binding_site_list <- list()

for ( i in regulator_names ) {
  file_names_i <- grep(i,file_names,value = T)
  file_names_i <- paste0("./regulator/",file_names_i)
  sub_list_i <- lapply(file_names_i, readRDS)
  combined_i <- Reduce(function(x,y)c(x,y), sub_list_i) %>% reduce
  binding_site_list[[i]] <- combined_i 
}

saveRDS(binding_site_list, "binding_site_list.rds")

rm(file_names_i, sub_list_i, combined_i, file_names)

assay_matrix <- matrix(0,nrow = length(rowRanges(m6Ads)), 
                         ncol = length(regulator_names))

colnames(assay_matrix) <- c(regulator_names)

for (i in regulator_names){
  assay_matrix[,i] <- rowRanges(m6Ads) %over% binding_site_list[[i]]
}

metadata(m6Ads)[["Regulators"]] <- assay_matrix
```


###Save on disk
```{r}
saveRDS(m6Ads, "../SE_full_multi.rds")
```

